File: BinarySearchTree/Populating NextRightPointersinEachNode.py
#LEETCODE SOLUTION

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root: return
        
        start = curr = root
        while curr.left:
            curr.left.next = curr.right
            if curr.next:
                curr.right.next = curr.next.left
                curr = curr.next
            else:
                curr.next = None
                curr = start.left
                start = curr
                
        return root
      
 #PEPCODING SOLUTION     
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        black=root
        while black is not None and black.left is not None:
            n=black
            while True:
                n.left.next=n.right
                if n.next is None:
                    break
                n.right.next=n.next.left
                n=n.next
            black= black.left
        return root
        

File: BinarySearchTree/BST_Java/Populating NextRightPointersinEachNode.java
//sol:https://www.youtube.com/watch?v=3MRPQFUpoA0
//https://leetcode.com/submissions/detail/757937604/
//ide: https://leetcode.com/problems/populating-next-right-pointers-in-each-node/

//PERFECT BINARY TREE: https://www.programiz.com/dsa/perfect-binary-tree


//TC: O(N)
//SC:O(N)

/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root==null) return null;
        
        Node start=root;
        Node curr=root;
        
        while(curr.left!=null){
            
            curr.left.next=curr.right;
            if(curr.next!=null){
                curr.right.next=curr.next.left;
                curr=curr.next;
                
            }
            
            else{
                curr.next=null;
                curr = start.left;
                start=curr;
            }
        }
        
        return root;
        
    }
}

